USIMD: Optimize the performace of np.einsum for all platforms

Qiyu8 opened the pull request:
seven sub module of np.einsum is optimized using neon intrinsics. The optimization resulted in a performance improvement of 10~42%.

### System info:

- HardWare: KunPeng
- Processor: ARMv8 2.6GMHZ 8 processors
- OS: Linux ecs-9d50 4.19.36-vhulk1905.1.0.h276.eulerosv2r8.aarch64

### The optimization effect:

einsum sub module triggered | master | einsum-neon | ratio
-- | -- | -- | --
sum_of_products_stride0_contig_outstride0_two | 110±0.5μs | 100±0.5μs | 0.90
sum_of_products_contig_stride0_outstride0_two | 110±0.5μs | 101±0.5μs | 0.90
sum_of_products_contig_stride0_outcontig_two | 23.7±1ms | 20.2±0.3ms | 0.85
sum_of_products_contig_two | 152±0.8μs | 140±2μs | 0.91
sum_of_products_stride0_contig_outcontig_two | 946±10μs | 841±20μs | 0.89
contig_contig_outstride0_two | 522±5μs | 353±7μs | 0.68
sum_of_products_contig_outstride0_one | 871±10μs | 502±4μs | 0.58

takanori-pskq committed:
DOC: Fixup

Qiyu8 committed:
optimize sum_of_products_stride0_contig_outcontig_two by using neon intrinsics

Qiyu8 committed:
optimize sum_of_products_stride0_contig_outcontig_two by using neon intrinsics

Qiyu8 committed:
optimize sum_of_products_contig_contig_outstride0_two by using neon intrinsics

Qiyu8 committed:
optimize sum_of_products_contig_contig_outstride0_two by using neon intrinsics

Qiyu8 committed:
optimize sum_of_products_contig_outstride0_one by using neon intrinsics

Qiyu8 committed:
optimize sum_of_products_contig_outstride0_one by using neon intrinsics

Qiyu8 committed:
add benchmarks

Qiyu8 committed:
add benchmarks

Qiyu8 committed:
optimize sum_of_products_contig_two using neon.

Qiyu8 committed:
optimize sum_of_products_contig_two using neon.

Qiyu8 committed:
optimize sum_of_products_contig_stride0_outcontig_two using neon.

Qiyu8 committed:
optimize sum_of_products_contig_stride0_outcontig_two using neon.

Qiyu8 committed:
optimize sum_of_products_contig_stride0_outcontig_two using neon.

Qiyu8 committed:
optimize sum_of_products_stride0_contig_outstride0_two using neon.

Qiyu8 committed:
optimize sum_of_products_stride0_contig_outstride0_two using neon.

Qiyu8 committed:
optimize sum_of_products_stride0_contig_outstride0_two using neon.

Qiyu8 committed:
optimize sum_of_products_contig_stride0_outstride0_two using neon.

Qiyu8 committed:
optimize sum_of_products_contig_stride0_outstride0_two using neon.

Qiyu8 committed:
optimize sum_of_products_contig_stride0_outstride0_two using neon.

Qiyu8 committed:
add dtype parameter

Qiyu8 committed:
add dtype parameter

Qiyu8 committed:
add dtype parameter

Commit:
ENH: Use f90 compiler specified in command line args for pgi compilers. Issue #16481

Qiyu8 committed:
rebase
Merge branch 'master' of https://github.com/numpy/numpy into einsum-neon

Qiyu8 committed:
modified accoriding to new NPY_HAVE_NEON flag.

Qiyu8 committed:
modified accoriding to new NPY_HAVE_NEON flag.

Qiyu8 committed:
modified accoriding to new NPY_HAVE_NEON flag.

seberg committed:
MAINT: Explicitly disallow object user dtypes

These never worked to the best of my knowledge, so disable
it explicitly to hopefully simplify cleanup in other parts.

mattip commented:
We discussed this briefly in the sprint. Correct me if I am wrong: this is a reimplementation of the SSE2 intrinsics in NEON. I wonder if we could already use the universal intrinsics to make this generic?

cjblocker committed:
BUG: fix mgrid output for lower precision float inputs

Floats besides float64 were being coerced to integers
and complex step sizes for the index trick classes
would fail for complex64 input. Fixes #16466

cjblocker committed:
TST: fixed dtype check error from code review

Qiyu8 commented:
I'd be happy to change it to universal intrinsics，Is there any guidance document or example that I should read first?

seiko2plus commented on numpy/core/src/multiarray/einsum.c.src:
```diff
@@ -37,6 +37,17 @@
 #define EINSUM_USE_SSE2 0
 #endif
 
+#ifdef NPY_HAVE_NEON
```
Use the following guards instead of CPU feature definitions.
```C
#if NPY_SIMD
// a valid guard for all data types except for float64 
#endif

#if NPY_SIMD_F64
// a special guard for float64, to add combitablty with arm7/neon
#endif
````

seiko2plus commented on numpy/core/src/multiarray/einsum.c.src:
```diff
@@ -45,6 +56,21 @@
 
+#if EINSUM_USE_NEON_32 || EINSUM_USE_NEON_64
+#include <arm_neon.h>
+static void prefetch(const double *ptr, size_t offset)
```
you may need to add a new intrin for prefetch or you can keep it as-is,
discover the SIMD extensions [dir](https://github.com/numpy/numpy/tree/master/numpy/core/src/common/simd) to see how we do it.

Qiyu8 committed:
rebase
Merge branch 'master' of https://github.com/numpy/numpy into einsum-neon

seiko2plus commented on numpy/core/src/multiarray/einsum.c.src:
```diff
@@ -275,6 +301,8 @@ static void
 #elif EINSUM_USE_SSE2 && @float64@
     __m128d a, b;
+#elif EINSUM_USE_NEON_64 && @float64@
+    float64x2_t a, b;
```
The new types are defined by the main extension header, e.g. [neon/neon.h](https://github.com/numpy/numpy/blob/master/numpy/core/src/common/simd/neon/neon.h).
however, in this case __m128d, and float64x2_t equivalent to npyv_f64.

seiko2plus commented on numpy/core/src/multiarray/einsum.c.src:
```diff
@@ -341,6 +369,25 @@ finish_after_unrolled_loop:
     }
+#elif EINSUM_USE_NEON_64 && @float64@
+        /* Unroll the loop by 8 */
+        while (count >= 8) {
```
Since in NPYV, we're dealing with different SIMD width, you can replace `8` with `npyv_nlanes_f64 * 4` where npyv_nlanes_{suffix type} is representing
the number of vector lanes.

*Note*: that may lead to a negative impact on avx512 since you're
unrolling a big chunk of memory.

However,  you can just unroll by 2 vectors on avx512, for example:
```C
#if NPY_SIMD_WIDTH > 32 # avx512f
const int unroll_by = npyv_nlanes_f64 * 2;
#else
const int unroll_by = npyv_nlanes_f64 * 4;
#endif
while (count >= unroll_by) {
  // ...
}
```

seiko2plus commented on numpy/core/src/multiarray/einsum.c.src:
```diff
@@ -341,6 +369,25 @@ finish_after_unrolled_loop:
+/**begin repeat2
+ * #i = 0, 2, 4, 6#
+ */
+            a = vmulq_f64(vld1q_f64(data0+@i@), vld1q_f64(data1+@i@));
```
replace these intrinsics with:
`vld1q_f64` -> `npyv_load_f64`
`vmulq_f64` -> `npyv_mul_f64`
`vaddq_f64` -> `npyv_add_f64`
`vst1q_f64` -> `npyv_store_f64`

for more intrinsics, you have to search or dive in the [NPYV](https://github.com/numpy/numpy/tree/master/numpy/core/src/common/simd)

seiko2plus replied:
BTW why you didn't use FMA3 here, well currently there's no mul-add intrinsic in NPYV but I guess It will not be that hard to impmlent it.

Qiyu8 replied:
I will make another PR for mul-add intrinsic in NPYV.

Qiyu8 committed:
Merge branch 'einsum-neon' of github.com:Qiyu8/numpy; branch 'master' of github.com:numpy/numpy into einsum-neon

cjblocker committed:
TST: update mgrid test from code review

Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>

cjblocker committed:
MAINT: reference issue in comments for added index_tricks tests

Qiyu8 committed:
recontructing einsum using usimd

Qiyu8 committed:
recontructing einsum using usimd

Qiyu8 committed:
recontructing einsum using usimd

Qiyu8 commented:
@seiko2plus Thanks for your advise, Reconstruct `einsum `using `SIMD `after reading `simd-optimizations.rst`, here is the routine:
1- Configuration: add a dispatch file to corresponding module that needs to optimized, remember to setup for compiler purpose.
2- Rewrite: rewrite the original platform related intrinsics by using `NPY_SIMD `and npyv simd api 
3- Test Case: write a SIMD dispatch test case to the caller module. in this case it's `multiarraymodule.c`, then add to `test_cpu_dispatcher` in order to trigger this test case.

I just make `@name@_sum_of_products_contig_two` function an example, correct me if i were wrong, then I will rewrite other functions in this module.

PS: USIMD is really powerful, saved so much coding time. 😃

mattip commented on .gitignore:
```diff
@@ -166,9 +166,8 @@ numpy/core/src/umath/simd.inc
-numpy/core/src/umath/_umath_tests.dispatch.h
-numpy/core/src/umath/_umath_tests.dispatch.sse41.c
+numpy/core/src/*/*.dispatch.*.c
+numpy/core/src/*/*.dispatch.h
```
@seiko2plus shouldn't we be generating files in build/* rather than in the source directories, or is this similar to cython-generated files?

seiko2plus replied:
yes, should be generated into build/src.*, but maybe @Qiyu8 use `--inplace`.

Qiyu8 replied:
yes, I used `-i` option during build.

seiko2plus replied:
I guess you used inplace option `-i` because you received error like that
```Bash
: fatal error: einsum.dispatch.h: No such file or directory
```
well it's my bad, `CcompilerOpt` should include `build/src.*/numpy/core/src/multiarray` to the search dir, even if it wasn't added by 'core/setup.py'.

This issue should be fixed by #16858.

mattip replied:
#16858 is merged.

seiko2plus replied:
thank you @mattip.

Qiyu8 replied:
I think we should keep this ignore part, I build the source with `-i` option, and the `*.dispatch.*.c` will generated.

mattip replied:
By default people should not be using `-i` since it messes up the source directory, as you discovered. Why don't you want to use the build directory? Is there another problem that using `-i` fixes?

Qiyu8 replied:
build directory is good , no problem if I using `-i` option. will remove from ignore list.

mattip commented on benchmarks/benchmarks/bench_linalg.py:
```diff
@@ -112,11 +112,14 @@ class Einsum(Benchmark):
         self.c1 = np.arange(1200, dtype=dtype).reshape(30, 40)
+        self.c2 = np.arange(480000, dtype=dtype)
+        self.c3 = np.arange(600, dtype=dtype)
         self.d = np.arange(10000, dtype=dtype).reshape(10,100,10)
```
It would be good at some point to rename these with meaningful names. Is there a reason to have the different size variants? Perhaps a very small variant would be good as well.

Qiyu8 replied:
@mattip I found that meaningless naming is pretty common in every `bench `module, should we open a separate rename PR? I want to focus on the universal intrinsics transform process in this PR.

mattip replied:
ok. What about adding a very small variant as well?

Qiyu8 replied:
I doubt that small variant is not enough to evaluate the degree of optimization, I will test with small size and provide a report anyway.

mattip commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -3832,6 +3832,58 @@ test_interrupt(PyObject *NPY_UNUSED(self), PyObject *args)
+    Py_XDECREF(item);
+    Py_DECREF(dict);
+    return NULL;
+}
```
this should go to one of the test extension modules, not to the main one. `_umath_tests` would be a good place.

mattip reviewed (commented):
nice to see use of the universal intrinsics.

seiko2plus commented:
@Qiyu8, you should dispatch the tabs that hold the functions pointers within `PyArray_EinsteinSum()` and also there's no need to defining tests. could you please allow me to help you? I'm just gonna initialize the dispatching process for you and then you will handle the universal intrinsics part.

Qiyu8 commented:
@seiko2plus Any kind of help is appreciated.

Qiyu8 committed:
using usimd based on current framework

Qiyu8 committed:
using usimd based on current framework

Qiyu8 committed:
using usimd based on current framework

seiko2plus committed:
initialize the cpu dispatching of einsum

seiko2plus committed:
initialize the cpu dispatching of einsum

seiko2plus committed:
initialize the cpu dispatching of einsum

Qiyu8 committed:
add prefetch in memory

Qiyu8 committed:
add prefetch in memory

Qiyu8 committed:
add prefetch in memory

Qiyu8 committed:
add reverse usimd

Qiyu8 committed:
add reverse usimd

Qiyu8 committed:
add reverse usimd

Qiyu8 committed:
Merge pull request #1 from seiko2plus/einsum-neon-dispatch

initialize the cpu dispatching of einsum

Qiyu8 committed:
Merge branch 'einsum-neon' of github.com:Qiyu8/numpy into einsum-neon

Qiyu8 committed:
Merge branch 'master' of github.com:numpy/numpy into einsum-neon

Qiyu8 committed:
rewrite using simd api

Qiyu8 committed:
rewrite using simd api

Qiyu8 committed:
rewrite using simd api

Qiyu8 committed:
Update numpy/core/src/common/simd/avx2/reorder.h

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/common/simd/avx2/reorder.h

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/common/simd/avx2/reorder.h

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/common/simd/avx512/reorder.h

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/common/simd/avx512/reorder.h

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/common/simd/avx512/reorder.h

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/common/simd/vsx/reorder.h

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/common/simd/vsx/reorder.h

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/common/simd/vsx/reorder.h

Co-authored-by: Sayed Adel <seiko@imavr.com>

Zac-HD committed:
Configure hypothesis for np.test()

Qiyu8 committed:
Merge branch 'master' of github.com:numpy/numpy into improve-usimd

Qiyu8 committed:
add shuffle api

Qiyu8 committed:
add shuffle api

Qiyu8 committed:
add shuffle api

Qiyu8 committed:
remove tabs and offset

Qiyu8 committed:
remove tabs and offset

Qiyu8 committed:
remove tabs and offset

takanori-pskq committed:
DOC: Fix wrong markups in `arrays.dtypes`

takanori-pskq committed:
DOC: Remove links for C codes

Ashutosh619-sudo committed:
Fix exception causes in __init__.py

Ashutosh619-sudo committed:
Fix exception causes in __init__.py

Qiyu8 committed:
Update numpy/core/src/common/simd/neon/reorder.h

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

Qiyu8 committed:
Update numpy/core/src/common/simd/neon/reorder.h

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

Qiyu8 committed:
Update numpy/core/src/common/simd/neon/reorder.h

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

Qiyu8 committed:
Update numpy/core/src/common/simd/neon/reorder.h

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

Qiyu8 committed:
Update numpy/core/src/common/simd/neon/reorder.h

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

Qiyu8 committed:
Update numpy/core/src/common/simd/neon/reorder.h

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

Qiyu8 committed:
Update numpy/core/src/common/simd/neon/reorder.h

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

Qiyu8 committed:
Update numpy/core/src/common/simd/neon/reorder.h

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

Qiyu8 committed:
Update numpy/core/src/common/simd/neon/reorder.h

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

Qiyu8 committed:
Merge branch 'improve-usimd' of github.com:Qiyu8/numpy into einsum-neon

Qiyu8 committed:
add shuffle api

Qiyu8 committed:
add shuffle api

Qiyu8 committed:
add shuffle api

Qiyu8 committed:
Merge branch 'master' of github.com:numpy/numpy into einsum-neon

Qiyu8 committed:
remove redundant func

Qiyu8 committed:
remove redundant func

Qiyu8 committed:
remove redundant func

Qiyu8 committed:
Transform to usimd, SSE/SSE2/AVX2 passed

Qiyu8 committed:
Transform to usimd, SSE/SSE2/AVX2 passed

Qiyu8 committed:
Transform to usimd, SSE/SSE2/AVX2 passed

Qiyu8 committed:
update
Merge branch 'einsum-neon' of github.com:Qiyu8/numpy into einsum-neon

Qiyu8 committed:
Merge branch 'master' of https://github.com/numpy/numpy into einsum-neon

Qiyu8 committed:
modify neon shuffle

Qiyu8 committed:
modify neon shuffle

Qiyu8 committed:
modify neon shuffle

Qiyu8 commented:
@seiko2plus The `einsum.dispatch.avx2.c` successfully generated in X86 machine, But `einsum.dispatch.asimddp.c` failed to generate in ARM machine, here is the log:
```
########### EXT COMPILER OPTIMIZATION ###########
CPU baseline  :
  Requested   : 'min'
  Enabled     : NEON NEON_FP16 NEON_VFPV4 ASIMD
  Flags       : none

CPU dispatch  :
  Requested   : 'asimddp'
  Enabled     : none
  Generated   : none
CCompilerOpt._cache_write[784] : write cache to path -> /usr/chunlin/numpy/build/temp.linux-aarch64-3.7/ccompiler_opt_cache_ext.py

########### CLIB COMPILER OPTIMIZATION ###########
CPU baseline  :
  Requested   : 'min'
  Enabled     : NEON NEON_FP16 NEON_VFPV4 ASIMD
  Flags       : none

CPU dispatch  :
  Requested   : 'max -xop -fma4'
  Enabled     : ASIMDHP
  Generated   : none
CCompilerOpt._cache_write[784] : write cache to path -> /usr/chunlin/numpy/build/temp.linux-aarch64-3.7/ccompiler_opt_cache_clib.py
```

eric-wieser commented on numpy/core/src/common/simd/neon/reorder.h:
```diff
@@ -53,6 +53,53 @@
+    return vreinterpretq_m128_f32(ret);
+}
+
+// shuffle vector lanes
```
Can you add a comment indicating what the identity shuffle is?

eric-wieser commented on numpy/core/src/common/simd/neon/reorder.h:
```diff
@@ -53,6 +53,53 @@
+    return ret;
+}
+#ifdef __aarch64__
+    // in fact it's revert operation
```
Not sure what this comment means.

eric-wieser commented on numpy/core/src/multiarray/einsum.dispatch.c.src:
```diff
@@ -0,0 +1,1866 @@
+        while (count >= UNROLL_BY) {
+            count -= UNROLL_BY;
+
+/**begin repeat2
+ * #i = 0, 1#
+ */
+            pack_size = @i@ * UNROLL_BY / 2;
+            a = npyv_mul_f32(value0_sse, npyv_loada_f32(data1+pack_size));
+            b = npyv_add_f32(a, npyv_loada_f32(data_out+pack_size));
+            npyv_storea_f32(data_out+pack_size, b);
+/**end repeat2**/
```
can you just use a regular `for` loop here? I don't expect the compiler to have any trouble unrolling it, and it makes things easier to read.

Qiyu8 replied:
good point, I will transform to regular `for` loop.

eric-wieser commented on numpy/core/src/multiarray/einsum.dispatch.c.src:
```diff
@@ -0,0 +1,1866 @@
+    }
+
+#if NPY_SIMD && @float32@
+    /* Add the four SSE values and put in accum */
+    a = npyv_shuffle_f32(accum_sse, accum_sse, _MM_SHUFFLE(2,3,0,1));
+    accum_sse = npyv_add_f32(a, accum_sse);
+    a = npyv_shuffle_f32(accum_sse, accum_sse, _MM_SHUFFLE(1,0,3,2));
+    accum_sse = npyv_add_f32(a, accum_sse);
+    npyv_storel_f32(&accum, accum_sse);
```
Is the end result of this shuffle to convert `[a, b, c ,d]` to `[a + b + c + d]*4`?

Qiyu8 replied:
@eric-wieser you are right, every lane of vector `accum_sse `is the sum of all lanes.

eric-wieser replied:
And then only one lane is actually used, right?

I think this would benefit from being extracted to a helper function, with an explanation of why such an odd approach is used.

Qiyu8 replied:
the odd approach has transformed to a more comprehended `sum `api.

Qiyu8 committed:
fix neon shuffle api

Qiyu8 committed:
fix neon shuffle api

Qiyu8 committed:
fix neon shuffle api

Qiyu8 committed:
Merge branch 'master' of https://github.com/numpy/numpy into einsum-neon

eric-wieser commented on numpy/core/src/common/simd/vsx/reorder.h:
```diff
@@ -31,6 +31,23 @@
+    // TODO: add simulation of shuffle
+    #define npyv_shuffle_f32 ((void)0)
+    #ifdef __clang__
+        NPY_FINLINE npyv_f64 npyv_shuffle_f64(npyv_f64 a, npyv_f64 b, int imm8)
+        { return vec_xxpermdi(a, b, 2); }
```
Why `2` and not `imm8`?

Qiyu8 replied:
The shuffle api  has removed.

eric-wieser commented on numpy/core/src/common/simd/vsx/reorder.h:
```diff
@@ -31,6 +31,23 @@
 #define npyv_combineh_f32(A, B) ((npyv_f32)npyv__combineh(A, B))
 #define npyv_combineh_f64       vec_mergel
 
+#if (defined(__GNUC__) && __GNUC__ >= 8) || (defined(__clang__) && __clang_major__ >= 6)
+    // TODO: add simulation of shuffle
+    #define npyv_shuffle_f32 ((void)0)
+    // https://www.ibm.com/support/knowledgecenter/SSLTBW_2.4.0/com.ibm.zos.v2r4.cbcpx01/bif_vec_reve.htm?view=embed
+    #define npyv_shuffle_f64 vec_reve
+#else
+    // missing on old versions of clang and gcc
+    // TODO: add simulation of shuffle
+    #define npyv_shuffle_f32 ((void)0)
+    #ifdef __clang__
+        NPY_FINLINE npyv_f64 npyv_shuffle_f64(npyv_f64 a, npyv_f64 b, int imm8)
+        { return vec_xxpermdi(a, b, 2); }
+    #else
+        NPY_FINLINE npyv_f64 npyv_shuffle_f64(npyv_f64 a, npyv_f64 b, int imm8)
+        { __asm__("xxswapd %x0,%x1" : "=wa" (a) : "wa" (b)); return a; }
+    #endif
+#endif
```
I'm very confuse by what you expect the signature of `npyv_shuffle_f64` to be - I count at least three different meanings, which vary by platform.

* `npyv_shuffle_f64(a)` - reverse `a`
* `npyv_shuffle_f64(a, b)` - do some kind of mix
* `npyv_shuffle_f64(a, b, imm8)` - do some kind of mixparametrized by imm8

mattip committed:
Merge pull request #16900 from Ashutosh619-sudo/master

MAINT: Chaining exceptions in numpy/__init__.py

mattip committed:
BLD: update OpenBLAS build

seberg committed:
BUG: Allow array-like types to be coerced as object array elements

This was previously allowed for nested cases, i.e.:

   np.array([np.int64])

but not for the scalar case:

   np.array(np.int64)

The solution is to align these two cases by always interpreting
these as not being array-likes (instead of invalid array-likes)
if the passed in object is a `type` object and the special
attribute has a `__get__` attribute (and is thus probable a
property or method).

The (arguably better) alterative to this is to move the special
attribute lookup to be on the type instead of the instance
(which is what python does). This will definitely require some
adjustments in our tests to use properties, but is probably
fine with respect to most actual code.
(The tests more commonly use this to quickly set up an
array-like, while it is a fairly strange pattern for typical code.)

Address parts of gh-16939
Closes gh-8877

charris committed:
Merge pull request #16940 from mattip/issue-16913

BLD: update OpenBLAS build

seberg committed:
DEP: Deprecate size-one ragged array coercion

Previously, code such as:
```
np.array([0, np.array([0])], dtype=np.int64)
```
worked by discovering the array as having a shape of `(2,)` and
then using `int(np.array([0]))` (which currently succeeds).
This is also a ragged array, and thus deprecated here earlier on.
(As a detail, in the new code the assignment should occur as
an array assignment and not using the `__int__`/`__float__`
Python protocols.)

Two details to note:

1. This still skips the deprecation for sequences which are not
   array-likes.  We assume that sequences will always fail the
   conversion to a number.
2. The conversion to a number (or string, etc.) may still fail
   after the DeprecationWarning is given.  This is not ideal, but
   narrowing it down seems tedious, since the actual assignment
   happens in a later step.
   I.e. `np.array([0, np.array([None])], dtype=np.int64)` will give
   the warning, but then fail anyway, since the array cannot be
   assigned.

Closes gh-16939

seberg committed:
Update numpy/core/tests/test_array_coercion.py

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

seberg committed:
Update numpy/core/src/multiarray/array_coercion.c

cjblocker committed:
DOC: add release note for #16815

takanori-pskq committed:
DOC: Fix the role of references (var -> macro)

Commit:
changed the name of the folder icons to logo

Qiyu8 committed:
add sum api to usimd

Qiyu8 committed:
add sum api to usimd

Qiyu8 committed:
add sum api to usimd

Qiyu8 committed:
remove print

Qiyu8 committed:
remove print

Qiyu8 committed:
remove print

rgommers committed:
Merge pull request #16944 from InessaPawson/master

MAINT: change the name of the folder "icons" to "logo".

mattip committed:
Merge pull request #16815 from cjblocker/mgrid-float

BUG: fix mgrid output for lower precision float inputs

charris committed:
Merge pull request #16943 from seberg/deprecate-single-element-arraylike-coercion

DEP: Deprecate size-one ragged array coercion

person142 committed:
ENH: enable colors for `runtests.py --ipython`

Currently the IPython shell from runtests has no colors; enable them.

Qiyu8 committed:
fix neon sum api, use normal for loop

Qiyu8 committed:
fix neon sum api, use normal for loop

Qiyu8 committed:
fix neon sum api, use normal for loop

Qiyu8 committed:
remove log

Qiyu8 committed:
remove log

Qiyu8 committed:
remove log

Qiyu8 committed:
open maxop option

Qiyu8 committed:
open maxop option

Qiyu8 committed:
open maxop option

mattip committed:
BLD: add win32 pypy build

rgommers committed:
Merge pull request #16949 from person142/runtests-ipython-colors

ENH: enable colors for `runtests.py --ipython`

dependabot-preview[bot] committed:
MAINT: Bump hypothesis from 5.20.2 to 5.23.2

Bumps [hypothesis](https://github.com/HypothesisWorks/hypothesis) from 5.20.2 to 5.23.2.
- [Release notes](https://github.com/HypothesisWorks/hypothesis/releases)
- [Commits](https://github.com/HypothesisWorks/hypothesis/compare/hypothesis-python-5.20.2...hypothesis-python-5.23.2)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

eric-wieser commented on numpy/core/src/common/simd/avx512/arithmetic.h:
```diff
@@ -112,5 +112,61 @@ NPY_FINLINE __m512i npyv_mul_u8(__m512i a, __m512i b)
+
+NPY_FINLINE __m256 get_high_f32(__m512 a)
+{
+    return _mm256_castpd_ps(_mm512_extractf64x4_pd(_mm512_castps_pd(a),1));
```
Why not `_mm512_extractf32x4_ps`?

Qiyu8 replied:
you are right, `_mm512_extractf32x8_ps` is better.

eric-wieser commented:
What was the performance impact of switching from your shuffle approach to the new method of computing sum?

charris committed:
Merge pull request #16952 from numpy/dependabot/pip/hypothesis-5.23.2

MAINT: Bump hypothesis from 5.20.2 to 5.23.2

charris committed:
MAINT: Use arm64 instead of aarch64 on travis.

Travis didn't recognize aarch65 and ran amd_64 instead. This
also fixes a compiler warning that turned up on with the fix.

charris committed:
Merge pull request #16957 from charris/fix-arm64-warning

MAINT: Use arm64 instead of aarch64 on travisCI.

seiko2plus commented:
@Qiyu8,
> The einsum.dispatch.avx2.c successfully generated in X86 machine, But einsum.dispatch.asimddp.c failed to generate in ARM machine, here is the log:
CPU dispatch  :
  Requested   : 'asimddp'
  Enabled     : none

It seems to me that `asimddp` isn't enabled because the compiler doesn't support it. you check the failure message of cpu_asimddp.c from the build log.
However, there's no need for `asimddp` since you not gonna optimize integer operation, so you can remove it.

Qiyu8 commented:
> What was the performance impact of switching from your shuffle approach to the new method of computing sum?

we use shuffle to do the sum reduce operation before this PR, this is ok because SSE/SSE2 is bored to have the shuffle instrument, but when using universal simd, I have to considering the universality through all of the instruction set, in AVX2 and AVX512, shuffle is different from it's meaning in SSE/SSE2,  and in VSX/Neon, there are no shuffle instrument, we have to simulate by using more instrument sets, which may be less efficient than the C implement.(https://github.com/jratcliff63367/sse2neon/blob/83725268b2fd1855740f3dfda5aff8ebab357c6c/SSE2NEON.h#L660), so I give up shuffle as a universal api to implement sum reduce, and use the sum api instead. we can use the optimal strategy to implement sum api according to each platform's specialty.

Qiyu8 committed:
use more efficient instrument.

Qiyu8 committed:
use more efficient instrument.

Qiyu8 committed:
use more efficient instrument.

nomanarshad94 committed:
update numpy/lib/arraypad.py with appropriate chain exception (#16953)

* update numpy/lib/arraypad.py with appropriate chain exception

Co-authored-by: Noman Arshad <noman.arshad@northbaysolutions.com>

Qiyu8 committed:
add AVX512DQ compatibility

Qiyu8 committed:
add AVX512DQ compatibility

Qiyu8 committed:
add AVX512DQ compatibility

Qiyu8 committed:
Merge branch 'einsum-neon' of https://github.com/Qiyu8/numpy into einsum-neon

Qiyu8 committed:
fix avx512 segment fault problem

Qiyu8 committed:
fix avx512 segment fault problem

Qiyu8 committed:
fix avx512 segment fault problem

mattip committed:
TST: fix tests for windows + PyPy

charris committed:
Merge pull request #16941 from seberg/types-are-not-arraylikes

BUG: Allow array-like types to be coerced as object array elements

Qiyu8 commented:
The Segmentation fault in CI is a litter strange, It's was caused by function address access violation, The code in question is shown below:
```
sop = get_sum_of_products_function(nop,
                        NpyIter_GetDescrArray(iter)[0]->type_num,
                        NpyIter_GetDescrArray(iter)[0]->elsize,
                        fixed_strides);
```
I print the variable `sop `inside of `get_sum_of_products_function` and get a correct address(such as 000007FEEED70FA0), and I get the wrong address outside this function(FFFFFFFFEED70FA0), The first four bytes was transformed, since `get_sum_of_products_function` is a dispatch method, @seiko2plus @mattip Any thoughts?

seiko2plus commented on numpy/core/src/multiarray/einsum_p.h:
```diff
@@ -0,0 +1,48 @@
+#include "common.h"
+#include "ctors.h"
+
+#define EINSUM_IS_ALIGNED(x) ((((npy_intp)x)&0xf) == 0)
```
```suggestion
#define EINSUM_IS_ALIGNED(x) npy_is_aligned(x, NPY_SIMD_WIDTH)
```
That one can easily causes a crush.

jesseli2002 committed:
DOC: Fixed typo in lib/recfunctions.py (#16973)

bharatr21 committed:
DOC: Clarify input to irfft/irfft2/irfftn (#16950)

* DOC: Clarify input to irfft/irfft2/irfftn
* DOC: Update see also link with rfft

seiko2plus committed:
re-implment SIMD kernels of einsum

  - paves the way for integer optimization
  - fix memory overflow/bus errors
  - improve the unrolling
  - activate the unrolling when NPYV isn't available
  - add support for fma3
  - robust/cleanup
  - other minor fixes

seiko2plus committed:
re-implment SIMD kernels of einsum

  - paves the way for integer optimization
  - fix memory overflow/bus errors
  - improve the unrolling
  - activate the unrolling when NPYV isn't available
  - add support for fma3
  - robust/cleanup
  - other minor fixes

seiko2plus committed:
re-implment SIMD kernels of einsum

  - paves the way for integer optimization
  - fix memory overflow/bus errors
  - improve the unrolling
  - activate the unrolling when NPYV isn't available
  - add support for fma3
  - robust/cleanup
  - other minor fixes

seiko2plus committed:
re-implment SIMD kernels of einsum

  - paves the way for integer optimization
  - fix memory overflow/bus errors
  - improve the unrolling
  - activate the unrolling when NPYV isn't available
  - add support for fma3
  - robust/cleanup
  - other minor fixes

seiko2plus committed:
re-implment SIMD kernels of einsum

  - paves the way for integer optimization
  - fix memory overflow/bus errors
  - improve the unrolling
  - activate the unrolling when NPYV isn't available
  - add support for fma3
  - robust/cleanup
  - other minor fixes

seiko2plus committed:
re-implment SIMD kernels of einsum

  - paves the way for integer optimization
  - fix memory overflow/bus errors
  - improve the unrolling
  - activate the unrolling when NPYV isn't available
  - add support for fma3
  - robust/cleanup
  - other minor fixes

seiko2plus reviewed (changes requested):
The following patch containing massive changes to the SIMD kernels that fix bus/overflow errors and also add many improvements.
Note: This patch removes prefetching read hint from the kernels since I don't think it has that much effect on modern compilers and hardware, however, it would be great if you test it (with/without).

[0001-re-implment-SIMD-kernels-of-einsum.patch.txt](https://github.com/numpy/numpy/files/4997792/0001-re-implment-SIMD-kernels-of-einsum.patch.txt)

charris committed:
Merge pull request #16974 from mattip/pypy-win32

TST: Add pypy win32 CI testing.

Qiyu8 committed:
Update numpy/core/src/multiarray/einsum_p.h

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/multiarray/einsum_p.h

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/multiarray/einsum_p.h

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
move to NPYV

Qiyu8 committed:
move to NPYV

Qiyu8 committed:
move to NPYV

Qiyu8 committed:
fix type error

Qiyu8 committed:
fix type error

Qiyu8 committed:
fix type error

BvB93 committed:
MAINT: Added the `order` parameter to `np.array()`  (#16966)

* MAINT: Added the `order` argument to `np.array()`

* MAINT: Made a number of `np.array()` arguments keyword-only

* TST: Added a `np.array` test with >2 positional arguments

Qiyu8 commented:
@seiko2plus Thanks for the Big improvements, I tested the performance with/without prefetch, the result is as follows:

einsum Benchmark | master | einsum-neon(without prefetch) | ratio | einsum-neon(with prefetch) | ratio
-- | -- | -- | -- | -- | --
time_einsum_contig_contig | 178±10μs | 107±5μs | 0.60 | 119±4μs | 0.66
time_einsum_contig_outstride0 | 313±10μs | 165±30μs | 0.53 | 183±70μs | 0.58
time_einsum_mul | 2.24±0.2ms | 2.18±0.2ms | 0.97 | 2.21±0.4ms | 0.98
time_einsum_multiply | 69.0±3μs | 63.9±4μs | 0.91 | 66.1±4μs | 0.95
time_einsum_outer | 46.2±2ms | 44.2±0.7ms | 0.95 | 45.2±0.5ms | 0.97
time_einsum_sum_mul | 56.7±1μs | 51.1±7μs | 0.91 | 55.9±4μs | 0.98
time_einsum_sum_mul2 | 57.0±1μs | 51.3±2μs | 0.91 | 53.0±2μs | 0.92

It's looks like the performance is not increased with prefetch, so let's remove it.  Next things todo:

- [ ] VSX sum reduce

- [ ] codecov

charris committed:
Merge pull request #16730 from danbeibei/fcompiler

ENH: Use f90 compiler specified in f2py command line args for pgi compiler

BvB93 committed:
MAINT: Implemented two dtype-related TODO's (#16622)

This pull requests addresses and implement two previous TODO's in `np.typing._DTypeLike`:
* Added a protocol, `_SupportsDtype`, for objects with the `dtype` attribute.
* Replaced an old dictionary, representing array field parameter, with a `TypedDict`.

DumbMice committed:
ENH: Add Neon SIMD implementations for add, sub, mul, and div (#16969)

* ENH: Add universal SIMD implementations of add, sub, mul, div for non-X86

leeyspaul committed:
DOC: update val to be scalar or array like optional closes #16901 (#16907)

* DOC: update val to be scalar or array like optional closes #16901

Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>

takanori-pskq committed:
DOC: Fix the declarations of C fuctions (#16897)

* DOC: Fix the declarations of C fuctions

mattip committed:
Merge pull request #16896 from takanori-pskq/i13114-5

DOC: Improve documentation of C code

takanori-pskq committed:
DOC: Remove the links for ``True`` and ``False`` (#16887)

* DOC: use :py:data:True and :py:data:False

Qiyu8 committed:
add vsx sum reduce

Qiyu8 committed:
add vsx sum reduce

Qiyu8 committed:
add vsx sum reduce

Qiyu8 commented:
@mattip I've never come across the codecov failure before, which detects 10 coverage changes, How can I fix this?

mattip commented:
I don't think the coverage report is interesting in this case. We do not aggregate results across platforms and architectures, so code for sse2 (32-bit x86) or arm64 or ppc64le will not be counted, lowering the coverage.

seiko2plus commented on numpy/core/src/common/simd/avx512/arithmetic.h:
```diff
@@ -112,5 +115,64 @@ NPY_FINLINE __m512i npyv_mul_u8(__m512i a, __m512i b)
+{
+    return _mm512_castps512_ps256(a);
+}
+
+NPY_FINLINE __m256 get_high_f32(__m512 a)
+{
+#ifdef NPY_HAVE_AVX512DQ
+    return _mm512_extractf32x8_ps(a, 1);
+#endif
+    return _mm256_castpd_ps(_mm512_extractf64x4_pd(_mm512_castps_pd(a),1));
+}
+
+NPY_FINLINE __m256d get_low_f64(__m512d a)
+{
+    return _mm512_castpd512_pd256(a);
+}
+
+NPY_FINLINE __m256d get_high_f64(__m512d a)
+{
+    return _mm512_extractf64x4_pd(a, 1);
+}
+
+// Horizontal add: Calculates the sum of all vector elements.
+NPY_FINLINE float npyv_sum_f32(__m512 a)
+{
+#if defined(__INTEL_COMPILER)
+    return _mm512_reduce_add_ps(a);
+#else
+    return npyv__sum_f32(_mm256_add_ps(get_low_f32(a), get_high_f32(a)));
+#endif
+}
+
+NPY_FINLINE double npyv_sum_f64(__m512d a)
+{
+#if defined(__INTEL_COMPILER)
+    return _mm512_reduce_add_pd(a);
+#else
+    return npyv__sum_f64(_mm256_add_pd(get_low_f64(a), get_high_f64(a)));
+#endif
+}
```
```suggestion
NPY_FINLINE float npyv_sum_f32(npyv_f32 a)
{
    __m512 h64   = _mm512_shuffle_f32x4(a, a, _MM_SHUFFLE(3, 2, 3, 2));
    __m512 sum32 = _mm512_add_ps(a, h64);
    __m512 h32   = _mm512_shuffle_f32x4(sum32, sum32, _MM_SHUFFLE(1, 0, 3, 2));
    __m512 sum16 = _mm512_add_ps(sum32, h32);
    __m512 h16   = _mm512_permute_ps(sum16, _MM_SHUFFLE(1, 0, 3, 2));
    __m512 sum8  = _mm512_add_ps(sum16, h16);
    __m512 h4    = _mm512_permute_ps(sum8, _MM_SHUFFLE(2, 3, 0, 1));
    __m512 sum4  = _mm512_add_ps(sum8, h4);
    return _mm512_cvtss_f32(sum4);
}
NPY_FINLINE double npyv_sum_f64(npyv_f64 a)
{
    __m512d h64   = _mm512_shuffle_f64x2(a, a, _MM_SHUFFLE(3, 2, 3, 2));
    __m512d sum32 = _mm512_add_pd(a, h64);
    __m512d h32   = _mm512_permutex_pd(sum32, _MM_SHUFFLE(1, 0, 3, 2));
    __m512d sum16 = _mm512_add_pd(sum32, h32);
    __m512d h16   = _mm512_permute_pd(sum16, _MM_SHUFFLE(2, 3, 0, 1));
    __m512d sum8  = _mm512_add_pd(sum16, h16);
    return _mm512_cvtsd_f64(sum8);
}
```

seiko2plus replied:
This implementation almost the same as `_mm512_reduce_add_ps/pd` but works all for all compilers.
I avoided the other technique that based on horizontally-add since `haddps/vhaddps/haddps/vhaddpd` became more expensive(latency) to use started from Skylake.

seiko2plus commented on numpy/core/src/common/simd/avx2/arithmetic.h:
```diff
@@ -62,6 +62,15 @@
+#ifdef NPY_HAVE_FMA3
+    #define npyv_muladd_f32 _mm256_fmadd_ps
+    #define npyv_muladd_f64 _mm256_fmadd_pd
+#endif
+
+#ifndef npyv_muladd_f32
```
```suggestion
#else
```

seiko2plus commented on numpy/core/src/common/simd/neon/arithmetic.h:
```diff
@@ -60,6 +60,12 @@
 #define npyv_mul_f32 vmulq_f32
 #define npyv_mul_f64 vmulq_f64
 
+#if NPYV_SIMD_F64 || defined(NPY_HAVE_NEON_VFPV4)
```
```suggestion
#ifdef NPY_HAVE_NEON_VFPV4
```

mattip committed:
Merge pull request #16894 from takanori-pskq/fix-doc-dtypes-quote

DOC: Fix wrong markups in `arrays.dtypes`

seiko2plus commented on numpy/core/src/common/simd/sse/arithmetic.h:
```diff
@@ -82,6 +82,15 @@ NPY_FINLINE __m128i npyv_mul_u8(__m128i a, __m128i b)
+#ifdef NPY_HAVE_FMA3
+    #define npyv_muladd_f32 _mm_fmadd_ps
+    #define npyv_muladd_f64 _mm_fmadd_pd
+#endif
+
+#ifndef npyv_muladd_f32
```
```suggestion
#else
```

seiko2plus commented on numpy/core/src/multiarray/einsum.dispatch.c.src:
```diff
@@ -0,0 +1,1522 @@
+    #endif
+    }
+    /**end repeat2**/
+#endif // NPYV check for @type@
```
```suggestion
npyv_cleanup();
#endif // NPYV check for @type@
```

seiko2plus commented on numpy/core/src/multiarray/einsum.dispatch.c.src:
```diff
@@ -0,0 +1,1522 @@
+    #endif
+    }
+    /**end repeat2**/
+#endif // NPYV check for @type@
```
```suggestion
npyv_cleanup();
#endif // NPYV check for @type@
```

seiko2plus commented on numpy/core/src/multiarray/einsum.dispatch.c.src:
```diff
@@ -0,0 +1,1522 @@
+    }
+    /**end repeat2**/
+    accum = npyv_sum_@sfx@(vaccum);
+#endif // NPYV check for @type@
```
```suggestion
npyv_cleanup();
#endif // NPYV check for @type@
```

seiko2plus commented on numpy/core/src/multiarray/einsum.dispatch.c.src:
```diff
@@ -0,0 +1,1522 @@
+    }
+    /**end repeat2**/
+    accum = npyv_sum_@sfx@(vaccum);
+#endif // NPYV check for @type@
```
```suggestion
npyv_cleanup();
#endif // NPYV check for @type@
```

seiko2plus commented on numpy/core/src/multiarray/einsum.dispatch.c.src:
```diff
@@ -0,0 +1,1522 @@
+    }
+    /**end repeat2**/
+    accum = npyv_sum_@sfx@(vaccum);
+#endif // NPYV check for @type@
```
```suggestion
npyv_cleanup();
#endif // NPYV check for @type@
```

seiko2plus commented on numpy/core/src/multiarray/einsum.dispatch.c.src:
```diff
@@ -0,0 +1,1522 @@
+    }
+    /**end repeat2**/
+    accum = npyv_sum_@sfx@(vaccum);
+#endif // NPYV check for @type@
```
```suggestion
npyv_cleanup();
#endif // NPYV check for @type@
```

bjnath committed:
DOC: Add the new NumPy logo to Sphinx pages

The changes are in one file local to us -- doesn't mess with any
scipy files.

bjnath committed:
DOC: Styling update for PR #16988

Incorporates suggestions by @seiko2plus

bjnath committed:
DOC: Delete old logo; updates PR #16988

rgommers committed:
Merge pull request #16988 from bjnath/update_logo_on_sphinx_pages

DOC: Add the new NumPy logo to Sphinx pages

mattip committed:
Merge pull request #16879 from Zac-HD/isolate-hypothesis-config

MAINT: Configure hypothesis in ``np.test()`` for determinism, use of global tmpdir

Qiyu8 committed:
Update numpy/core/src/common/simd/neon/arithmetic.h

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/common/simd/neon/arithmetic.h

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/common/simd/neon/arithmetic.h

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/common/simd/avx2/arithmetic.h

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/common/simd/avx2/arithmetic.h

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/common/simd/avx2/arithmetic.h

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/common/simd/avx512/arithmetic.h

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/common/simd/avx512/arithmetic.h

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/common/simd/avx512/arithmetic.h

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/common/simd/sse/arithmetic.h

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/common/simd/sse/arithmetic.h

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/common/simd/sse/arithmetic.h

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/multiarray/einsum.dispatch.c.src

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/multiarray/einsum.dispatch.c.src

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/multiarray/einsum.dispatch.c.src

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/multiarray/einsum.dispatch.c.src

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/multiarray/einsum.dispatch.c.src

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/multiarray/einsum.dispatch.c.src

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/multiarray/einsum.dispatch.c.src

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/multiarray/einsum.dispatch.c.src

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/multiarray/einsum.dispatch.c.src

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/multiarray/einsum.dispatch.c.src

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/multiarray/einsum.dispatch.c.src

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/multiarray/einsum.dispatch.c.src

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/multiarray/einsum.dispatch.c.src

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/multiarray/einsum.dispatch.c.src

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/multiarray/einsum.dispatch.c.src

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/multiarray/einsum.dispatch.c.src

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/multiarray/einsum.dispatch.c.src

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/multiarray/einsum.dispatch.c.src

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/multiarray/einsum.dispatch.c.src

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/multiarray/einsum.dispatch.c.src

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/multiarray/einsum.dispatch.c.src

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/multiarray/einsum.dispatch.c.src

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/multiarray/einsum.dispatch.c.src

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 committed:
Update numpy/core/src/multiarray/einsum.dispatch.c.src

Co-authored-by: Sayed Adel <seiko@imavr.com>

Qiyu8 commented:
The CI failure is the same with #16987 and #16960 , but no one modified `distutils/command/develop.py` recently.

rgommers commented:
@Qiyu8 when you making a set of changes in different commits at the same time, you may want to work locally for a little bit and then push them all at once. When you push 1 commit to your PR in a row, you trigger CI runs for each commit, and then abort them again when you push the next commit. We're close to our maximum CI capacity, so this can lead to delays for other PRs.

Qiyu8 commented:
@rgommers Thanks for explains the CI trigger process in detail, I'll do better next time.

rgommers commented:
Thanks, and no worries:)

mattip committed:
BLD: pin setuptools<49.2.0

charris committed:
Merge pull request #16993 from mattip/setuptools

BLD: pin setuptools<49.2.0

dependabot-preview[bot] committed:
MAINT: Bump hypothesis from 5.23.2 to 5.23.9

Bumps [hypothesis](https://github.com/HypothesisWorks/hypothesis) from 5.23.2 to 5.23.9.
- [Release notes](https://github.com/HypothesisWorks/hypothesis/releases)
- [Commits](https://github.com/HypothesisWorks/hypothesis/compare/hypothesis-python-5.23.2...hypothesis-python-5.23.9)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

anirudh2290 committed:
DOC: Add correctness vs strictness consideration for np.dtype (#16917)

* DOC: Add correctness vs strictness consideration for np.dtype

* MAINT: Update numpy/typing/__init__.py

Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>

* MAINT: Update numpy/typing/__init__.py

Co-authored-by: Bas van Beek <43369155+BvB93@users.noreply.github.com>

* MAINT: Change type to static type to differentiate from type objects

* MAINT: Revert static type change

Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>
Co-authored-by: Bas van Beek <43369155+BvB93@users.noreply.github.com>

abalkin committed:
BUG: Set readonly flag in array interface instead of warning (gh-16350)

Set readonly flags in array interface dict and array struct
capsule returned by an array with warn_on_write flag set
instead of producing a deprecation warning.

Closes #16335

charris committed:
Merge pull request #16991 from numpy/dependabot/pip/hypothesis-5.23.9

MAINT: Bump hypothesis from 5.23.2 to 5.23.9

dependabot-preview[bot] committed:
MAINT: Bump pytest from 5.4.3 to 6.0.1

Bumps [pytest](https://github.com/pytest-dev/pytest) from 5.4.3 to 6.0.1.
- [Release notes](https://github.com/pytest-dev/pytest/releases)
- [Changelog](https://github.com/pytest-dev/pytest/blob/master/CHANGELOG.rst)
- [Commits](https://github.com/pytest-dev/pytest/compare/5.4.3...6.0.1)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

mattip committed:
Merge pull request #16802 from seberg/user-dtypes-no-objects

MAINT: Explicitly disallow object user dtypes

charris committed:
Merge pull request #16992 from numpy/dependabot/pip/pytest-6.0.1

MAINT: Bump pytest from 5.4.3 to 6.0.1

BvB93 committed:
ENH: Speed up trim_zeros (#16911)

* Added a benchmark for `trim_zeros()`

* Improve the performance of `np.trim_zeros()`

* Increase the variety of the tests

Fall back to the old `np.trim_zeros()` implementation if an exception is encountered.
Emit a `DeprecationWarning` in such case.


* DEP,REL: Added a deprecation release note

nomanarshad94 committed:
MAINT: Chain exception in ``distutils/fcompiler/environment.py``. (#16962)

Fixup exception chaining and more detailed exception message

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>
Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>

iamsoto committed:
DOC: Add note that allclose and isclose do not accept non-numeric types (#16835)


Co-authored-by: Warren Weckesser <warren.weckesser@gmail.com>

scoder committed:
ENH: Add NumPy declarations to be used by Cython 3.0+ (#16986)

* Create copy of numpy.pxd for Cython 3.0 changes and improve it.

dkogan committed:
DOC: Improve intersect1d docstring (#16420)

The docstring now says what to expect if you call
intersect1d(assume_unique=True) but pass in non-unique data.

Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>

eric-wieser committed:
MAINT: Improve error handling in umathmodule setup (#17014)

Really this only ensures MemoryErrors propagate, nothing else can go wrong here.

phoenix-meadowlark committed:
DOC: Fix non-matching pronoun in format.py documentation. (gh-17022)

Follow up to [this commit](https://github.com/numpy/numpy/commit/36f71788620bb17d123669cf03f086fb2b87b3a8), which missed the second pronoun in this sentence.

asmeurer committed:
BUG: Raise correct errors in boolean indexing fast path (gh-17010)

Previously the logic assumed that an index was valid if the size was the same
as the indexed array, rather than the shape. This resulted in an index being
incorrectly allowed if the index was all False, like np.empty((3, 3))[np.full((1, 9),
False)], or incorrectly giving a ValueError about broadcasting if the index
was not all False (like np.empty((3, 3))[np.full((1, 9), True)]). Now these
examples both give IndexError with the usual error message.

Fixes #16997.

pentschev committed:
NEP: Updated NEP-35 with keyword-only instruction (#17009)

* NEP: Updated NEP-35 with keyword-only instruction

Following reviewer's suggestions during the implementation phase,
the `like=` argument is now keyword-only, updating the proposal
accordingly.

* NEP: Changed date of NEP-35 latest update

* NEP: Adjust keyword-only marker position in NEP-35

seiko2plus commented:
@Qiyu8, please could you rebase this patch into two commits, the first one representing the new npyv intrinsics and the other covering the other changes?
Also we need a fair benchmark for X86 to detect the performance changes before the cleanup review, by fair I mean a benchmark that based on the baseline build since the old code wasn't supporting runtime dispatching. For example:
```Bash
// Equal OPT level since policy `$maxopt` is been used within the configration statments of the dispatch-able source.
export CFLAGS="-O3"
// AVX512F
python setup.py build -j(`nproc`) --cpu-baseline="AVX512F" --cpu-dispatch="none" install
// AVX2 with FMA
python setup.py build -j(`nproc`) --cpu-baseline="AVX2 FMA3" --cpu-dispatch="none" install
// SSE2
python setup.py build -j(`nproc`) --cpu-baseline="SEE2" --cpu-dispatch="none" install
````
The idea behind this approach is to allow the compiler to replace the existence SSE2 with AVX2/AVX512F also to test the changes on non-AVX machines. From my side, I will provide benchmarks for Power8/9.

Another thing the benchmark should cover scalar/contiguous/reduce, feel free to assign the whole work to me in case if you face lack of time.

bjnath committed:
DOC: Disclaimer for FFT library

Adds sentence to the top of routines.fft.html explaining that scipy.fft
is a more comprehensive collection, per mailing list discussion
(http://numpy-discussion.10968.n7.nabble.com/Add-Chebyshev-cosine-transforms-implemented-via-FFTs-td48373.html#a48382)

mattip committed:
BUG: fix a compile and a test warning

rgommers committed:
Merge pull request #17028 from bjnath/fft-disclaimer

DOC: Disclaimer for FFT library

charris committed:
Merge pull request #17033 from mattip/random-pool_size

BUG: fix a compile and a test warning

yogeshr59 committed:
DOC: Use a less ambiguous example for array_split (#17039)

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

Qiyu8 committed:
Merge branch 'einsum-neon' of github.com:Qiyu8/numpy into einsum-neon

Qiyu8 committed:
Merge branch 'einsum-neon' of github.com:Qiyu8/numpy into einsum-neon
